kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/RecordConverter.java

```diff
@@ -0,0 +1,508 @@
+                    String fieldName = tableSchema.findColumnName(tableField.fieldId());
+                    schemaUpdateConsumer.makeOptional(fieldName);
+                    hasSchemaUpdates = true;
+                  }
+                }
+                if (!hasSchemaUpdates) {
+                  result.setField(
+                      tableField.name(),
+                      convertValue(
+                          struct.get(recordField),
+                          tableField.type(),
+                          tableField.fieldId(),
+                          schemaUpdateConsumer));
+                }
+              }
+            });
+    return result;
+  }
+
+  private NestedField lookupStructField(String fieldName, StructType schema, int structFieldId) {
+    if (nameMapping == null) {
+      return config.schemaCaseInsensitive()
+          ? schema.caseInsensitiveField(fieldName)
+          : schema.field(fieldName);
+    }
+
+    return structNameMap
+        .computeIfAbsent(structFieldId, notUsed -> createStructNameMap(schema))
+        .get(fieldName);
+  }
+
+  private Map<String, NestedField> createStructNameMap(StructType schema) {
+    Map<String, NestedField> map = Maps.newHashMap();
+    schema
+        .fields()
+        .forEach(
+            col -> {
+              MappedField mappedField = nameMapping.find(col.fieldId());
+              if (mappedField != null && !mappedField.names().isEmpty()) {
+                mappedField.names().forEach(name -> map.put(name, col));
+              } else {
+                map.put(col.name(), col);
+              }
+            });
+    return map;
+  }
+
+  protected List<Object> convertListValue(
+      Object value, ListType type, SchemaUpdate.Consumer schemaUpdateConsumer) {
+    Preconditions.checkArgument(value instanceof List);
+    List<?> list = (List<?>) value;
+    return list.stream()
+        .map(
+            element -> {
+              int fieldId = type.fields().get(0).fieldId();
+              return convertValue(element, type.elementType(), fieldId, schemaUpdateConsumer);
+            })
+        .collect(Collectors.toList());
+  }
+
+  protected Map<Object, Object> convertMapValue(
+      Object value, MapType type, SchemaUpdate.Consumer schemaUpdateConsumer) {
+    Preconditions.checkArgument(value instanceof Map);
+    Map<?, ?> map = (Map<?, ?>) value;
+    Map<Object, Object> result = Maps.newHashMap();
+    map.forEach(
+        (k, v) -> {
+          int keyFieldId = type.fields().get(0).fieldId();
+          int valueFieldId = type.fields().get(1).fieldId();
+          result.put(
+              convertValue(k, type.keyType(), keyFieldId, schemaUpdateConsumer),
+              convertValue(v, type.valueType(), valueFieldId, schemaUpdateConsumer));
+        });
+    return result;
+  }
+
+  protected int convertInt(Object value) {
+    if (value instanceof Number) {
+      return ((Number) value).intValue();
+    } else if (value instanceof String) {
```

Comment:
I've noticed that the String part for ints/longs/floats/doubles isn't tested in the existing tests. maybe it would be good to test those if that's easily possible

Reply:
Yes, I agree. I added test cases for these.